dpctl/tensor/libtensor/include/utils/sycl_utils.hpp

```diff
@@ -154,6 +154,30 @@ T custom_reduce_over_group(const GroupT &wg,
     return sycl::group_broadcast(wg, red_val_over_wg);
 }
 
+template <typename T, typename GroupT, typename LocAccT, typename OpT>
+T custom_inclusive_scan_over_group(const GroupT &wg,
+                                   LocAccT local_mem_acc,
+                                   const T local_val,
+                                   const OpT &op)
+{
+    auto local_id = wg.get_local_id(0);
+    auto wgs = wg.get_local_range(0);
+    local_mem_acc[local_id] = local_val;
+
+    sycl::group_barrier(wg, sycl::memory_scope::work_group);
+
+    if (wg.leader()) {
+        for (size_t i = 1; i < wgs; ++i) {
+            local_mem_acc[i] = op(local_mem_acc[i], local_mem_acc[i - 1]);
+        }
+    }
+
+    T accumulated_local_val = local_mem_acc[local_id];
+    sycl::group_barrier(wg, sycl::memory_scope::work_group);
```

Comment:
```suggestion
    if (wg.leader()) {
        T prev = local_mem_acc[0];
        for (size_t i = 1; i < wgs; ++i) {
            prev = op(local_mem_acc[i], prev]);
            local_mem_acc[i] = prev;
        }
    }
    
    // ensure all work-items see the same SLM that leader updated
    sycl::group_barrier(wg, sycl::memory_scope::work_group);
    T accumulated_local_val = local_mem_acc[local_id];
    // ensure all work-items updated the local variable
    sycl::group_barrier(wg, sycl::memory_scope::work_group);
```

Reply:
I feel like it is more appropriate to move the second `group_barrier` out of this function to after its call site.

Use of private variable `scan_val` helps do more work in registers, reducing bandwidth of SLM access,